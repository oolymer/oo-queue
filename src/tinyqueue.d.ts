declare module "tinyqueue" {

  export default class TinyQueue {
    constructor(data?: any[], compare?: Comparator)

    data: any[]
    length: number
    compare: Comparator

    push(item: any): void
    pop(): any
    peek(): any
  }

  type Comparator = (a: any, b: any) => number

}
