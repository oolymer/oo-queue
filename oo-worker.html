<link rel="import" href="../polymer/polymer-element.html">

<dom-module id="oo-worker">
  <template>
    <style>
      :host {
        display: block;
      }
    </style>

    worker
  </template>

  <script src="oo-worker.min.js"></script>

  <script>
    // Use local function scope to avoid bleeding to many variables into global `window`.
    (function () {
      // Declare or reuse global namespace.
      window.oo = window.oo || {}

      /**
       * @customElement
       * @polymer
       */
      oo.OoWorker = class OoWorker extends Polymer.Element {
        static get is() {
          return "oo-worker"
        }

        static get properties() {
          return {
          }
        }

        ready() {
          super.ready()
          console.log(oo)

          this._worker = new oo.Worker()
          console.log(this._worker)
        }
      }

      window.customElements.define(oo.OoWorker.is, oo.OoWorker)
    }())
  </script>
</dom-module>
