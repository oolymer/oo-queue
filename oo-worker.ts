import { queue, Queue } from "d3-queue"
import TinyQueue from "tinyqueue"

export { queue, TinyQueue }

export function worker() {
  const t = new TinyQueue()
  const q = queue()
  return [t, q]
}

export class Worker {

  tasks: TinyQueue
  queue: Queue

  constructor() {
    this.tasks = new TinyQueue()
    this.queue = queue()

    this.tasks.push("foo")
    this.tasks.push("bar")
    this.tasks.push("baz")
    this.tasks.push("quux")

    for (const task of this.tasks.data) {
      this.queue.defer(done => {
        console.log(task)
        done(null)
      })
    }

    // this.queue.abort()
    this.queue.awaitAll(error => {
      console.log(error)
    })
  }

}
