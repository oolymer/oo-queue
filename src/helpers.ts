export function _check(condition: boolean, lazyMessage?: () => string) {
  if (!condition) {
    if (lazyMessage) {
      throw new Error(lazyMessage())
    }
    throw new Error("illegal argument")
  }
}

export function _checkNotNil(value: any, lazyMessage?: () => string) {
  if (_isNil(value)) {
    if (lazyMessage) {
      throw new Error(lazyMessage())
    }
    throw new Error("illegal argument")
  }
}

export function _isNil(value: any): boolean {
  return value === null || value === undefined
}
