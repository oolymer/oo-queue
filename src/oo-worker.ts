/// <reference path="tinyqueue.d.ts" />

import { queue, Queue } from "d3-queue"
import TinyQueue from "tinyqueue"

export class Worker {
  _tasks: TinyQueue
  _queue: Queue

  constructor() {
    this._tasks = new TinyQueue()
    this._queue = queue()

    for (let index = 0; index < 25; index += 1) {
      this._tasks.push({ data: index + 1 })
    }

    for (const task of this._tasks.data) {
      this._queue.defer(done => {
        this._log(task)
        done(null)
      })
    }

    // this.queue.abort()
    this._queue.awaitAll(error => {
      this._log(error)
    })
  }

  _log(...args: any[]) {
    console.log(`[${this.constructor.name}]`, ...args)
  }
}
