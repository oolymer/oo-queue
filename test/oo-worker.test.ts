import { Worker } from "../src/oo-worker"

describe("oo-worker", () => {

  it("should contain tasks", () => {
    const worker = new Worker()
    expect(worker._tasks.data).toMatchSnapshot()
  })

})
