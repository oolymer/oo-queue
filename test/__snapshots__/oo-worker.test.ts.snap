// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`oo-worker should contain tasks 1`] = `
Array [
  Object {
    "data": 1,
  },
  Object {
    "data": 2,
  },
  Object {
    "data": 3,
  },
  Object {
    "data": 4,
  },
  Object {
    "data": 5,
  },
  Object {
    "data": 6,
  },
  Object {
    "data": 7,
  },
  Object {
    "data": 8,
  },
  Object {
    "data": 9,
  },
  Object {
    "data": 10,
  },
  Object {
    "data": 11,
  },
  Object {
    "data": 12,
  },
  Object {
    "data": 13,
  },
  Object {
    "data": 14,
  },
  Object {
    "data": 15,
  },
  Object {
    "data": 16,
  },
  Object {
    "data": 17,
  },
  Object {
    "data": 18,
  },
  Object {
    "data": 19,
  },
  Object {
    "data": 20,
  },
  Object {
    "data": 21,
  },
  Object {
    "data": 22,
  },
  Object {
    "data": 23,
  },
  Object {
    "data": 24,
  },
  Object {
    "data": 25,
  },
]
`;
